dhis-2/dhis-api/src/main/java/org/hisp/dhis/user/CurrentUserUtil.java

```diff
@@ -173,20 +176,96 @@ public static <T> T getUserSetting( UserSettingKey key )
     public static void initializeUser( User user )
     {
         initializeAndUnproxy( user );
+        initializeAndUnproxy( user.getPreviousPasswords() );
+        initializeAndUnproxy( user.getApps() );
         initializeAndUnproxy( user.getOrganisationUnits() );
+        initializeAndUnproxy( user.getTeiSearchOrganisationUnits() );
+        initializeAndUnproxy( user.getDataViewOrganisationUnits() );
+        user.getOrganisationUnits().stream().filter( Objects::nonNull )
+            .forEach( organisationUnit -> initializeAndUnproxy( organisationUnit.getChildren() ) );
+        initializeAndUnproxy( user.getTeiSearchOrganisationUnits() );
+        user.getTeiSearchOrganisationUnits().stream().filter( Objects::nonNull )
+            .forEach( organisationUnit -> initializeAndUnproxy( organisationUnit.getChildren() ) );
+        initializeAndUnproxy( user.getDataViewOrganisationUnits() );
+        user.getDataViewOrganisationUnits().stream().filter( Objects::nonNull )
+            .forEach( organisationUnit -> initializeAndUnproxy( organisationUnit.getChildren() ) );
+
         initializeAndUnproxy( user.getUserRoles() );
         for ( UserRole userRole : user.getUserRoles() )
         {
+            if ( userRole == null )
+            {
+                continue;
+            }
             initializeAndUnproxy( userRole );
             initializeAndUnproxy( userRole.getAuthorities() );
+            initializeAndUnproxy( userRole.getMembers() );
+            initializeAndUnproxy( userRole.getUserAccesses() );
+            initializeAndUnproxy( userRole.getAttributeValues() );
         }
+
         initializeAndUnproxy( user.getGroups() );
-        initializeAndUnproxy( user.getTeiSearchOrganisationUnits() );
-        initializeAndUnproxy( user.getDataViewOrganisationUnits() );
+        for ( UserGroup group : user.getGroups() )
+        {
+            if ( group == null )
+            {
+                continue;
+            }
+            initializeAndUnproxy( group.getManagedByGroups() );
+            initializeAndUnproxy( group.getManagedGroups() );
+            initializeAndUnproxy( group.getMembers() );
```

Comment:
This can potentially be very expensive. On some instances they can have many thousand users/members, and many hundreds of user groups, e.g. for PEPFAR. This will be very heavy to fetch on every login.

Reply:
Maybe we can do some tests on some DBs that is similar to get a idea how expensive it is. Another thing to consider might be to refactor the usage of user groups to be used through a service, so we do more efficient caching etc.